src/app/common/status-controller.component.ts

```diff
@@ -0,0 +1,81 @@
+import { Component, Input, NgModule } from "@angular/core";
+import { Status } from "./status";
+import { FormsModule } from "@angular/forms";
+import { CommonModule } from "@angular/common";
+
+export interface StatusModel {
+    readonly status: Status;
+}
+
+export abstract class StatusController {
+    private inProgress: boolean = false;
+    abstract StartImpl(): Promise<any>;
+    abstract StopImpl(): Promise<any>;
+    constructor(
+        private model: StatusModel,
+    ) {}
+
+    RestartImpl(): Promise<any> {
+        return this.StartImpl().then(this.StopImpl);
+    }
+
+    CanStart() {
+        return !this.inProgress && this.model.status == Status.Stopped;
+    }
+
+    Start() {
+        this.Invoke(this.StartImpl());
+    }
+    
+    CanStop() {
+        return !this.inProgress && this.model.status == Status.Started;
+    }
+
+    Stop() {
+        this.Invoke(this.StopImpl());
+    }
+
+    Restart() {
+        this.Invoke(this.RestartImpl());
+    }
+
+    Invoke(operation: Promise<any>) {
+        this.inProgress = true;
+        operation.finally(() => this.inProgress = false);
+    }
+}
+
+@Component({
```

Comment:
Please put @Component first before class. I think that is the coding convention.